### ISSUE TITLE
python: Segfault on exit with certain scripts

### ISSUE THREAD

**Issue by adsr** [weechat/weechat#2125]
### Describe the bug

The python plugin segfaults on exit with certain scripts on my setup.

### Steps to reproduce

Run `weechat -at` and the following commands:
```
/python load /path/to/scripts/go.py
/python load /path/to/scripts/mop.py
/quit
```

Strangely if I load `mop.py` before `go.py`, there's no segfault.

### Current behavior

Segfault at exit

### Expected behavior

No segfault

### Additional information

```
Linux - 6.8.0-31-generic #31-Ubuntu SMP PREEMPT_DYNAMIC Sat Apr 20 00:40:06 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
Python 3.12.3
WeeChat 4.3.1
```

Stacktrace and ASAN output below.

```
(gdb) bt full
#0  0x00007ffff5d25b31 in PyObject_GC_UnTrack () from /lib/x86_64-linux-gnu/libpython3.12.so.1.0
No symbol table info available.
#1  0x00007ffff5bdd1f2 in ?? () from /lib/x86_64-linux-gnu/libpython3.12.so.1.0
No symbol table info available.
#2  0x00007ffff5bdc6b5 in ?? () from /lib/x86_64-linux-gnu/libpython3.12.so.1.0
No symbol table info available.
#3  0x00007ffff5bce648 in ?? () from /lib/x86_64-linux-gnu/libpython3.12.so.1.0
No symbol table info available.
#4  0x00007ffff5bde692 in _PyModule_ClearDict () from /lib/x86_64-linux-gnu/libpython3.12.so.1.0
No symbol table info available.
#5  0x00007ffff5cef1e0 in ?? () from /lib/x86_64-linux-gnu/libpython3.12.so.1.0
No symbol table info available.
#6  0x00007ffff5cf2fee in Py_EndInterpreter () from /lib/x86_64-linux-gnu/libpython3.12.so.1.0
No symbol table info available.
#7  0x00007ffff63bc6a1 in weechat_python_unload (script=0x55555605d7d0) at /home/adam/weechat-src/src/plugins/python/weechat-python.c:944
        rc = 0x55555613d5b0
        interpreter = 0x7ffff48a1930
        filename = 0x555555fb4440 "/home/adam/.weechat/python/autoload/mop.py"
#8  0x00007ffff63bc862 in weechat_python_unload_all () at /home/adam/weechat-src/src/plugins/python/weechat-python.c:992
No locals.
#9  0x00007ffff63f16e2 in plugin_script_end (weechat_plugin=0x555555ca8960, plugin_data=0x7ffff63ff640 <python_data>) at /home/adam/weechat-src/src/plugins/plugin-script.c:1827
        scripts_loaded = 1
#10 0x00007ffff63be29f in weechat_plugin_end (plugin=0x555555ca8960) at /home/adam/weechat-src/src/plugins/python/weechat-python.c:1627
No locals.
#11 0x000055555563405f in plugin_unload (plugin=0x555555ca8960) at /home/adam/weechat-src/src/plugins/plugin.c:1254
        end_func = 0x7ffff63be23a <weechat_plugin_end>
        name = 0x555555cd06c0 "python"
#12 0x00005555556341a8 in plugin_unload_all () at /home/adam/weechat-src/src/plugins/plugin.c:1305
        plugins_loaded = 1
#13 0x00005555556344ee in plugin_end () at /home/adam/weechat-src/src/plugins/plugin.c:1425
No locals.
#14 0x0000555555563187 in weechat_end (gui_end_cb=0x55555562ace9 <gui_main_end>) at /home/adam/weechat-src/src/core/weechat.c:698
No locals.
#15 0x0000555555562137 in main (argc=2, argv=0x7fffffffde68) at /home/adam/weechat-src/src/gui/curses/normal/main.c:45
No locals.
```

```
=================================================================
==9602==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 262144 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fb4d0 in calloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:77
    #1 0x79600b9e5b3b  (<unknown module>)
    #2 0x79600b9e6819  (<unknown module>)
    #3 0x79600b9e6cfc  (<unknown module>)
    #4 0x79600bb25de5  (/home/adam/weechat/bin/weechat+0x4e0de5)
    #5 0x79600b9ca49b  (<unknown module>)
    #6 0x79600ba5146a  (/home/adam/weechat/bin/weechat+0x40c46a)
    #7 0x79600baefe59  (/home/adam/weechat/bin/weechat+0x4aae59)
    #8 0x79600baf1e9d  (/home/adam/weechat/bin/weechat+0x4ace9d)
    #9 0x79600baf2024  (/home/adam/weechat/bin/weechat+0x4ad024)
    #10 0x79600baf327f  (/home/adam/weechat/bin/weechat+0x4ae27f)
    #11 0x79600ef46d7b  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d7b)
    #12 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #13 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152
    #14 0x59a696cfa428 in plugin_init /home/adam/weechat-src/src/plugins/plugin.c:1409
    #15 0x59a696b1e004 in weechat_init /home/adam/weechat-src/src/core/weechat.c:675
    #16 0x59a696b1c55d in main /home/adam/weechat-src/src/gui/curses/normal/main.c:43
    #17 0x79601082a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #18 0x79601082a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #19 0x59a696b1c464 in _start (/home/adam/weechat/bin/weechat+0x9c464) (BuildId: 405e4255136b6fd0fefff49787a34a3e899b7b0c)

Direct leak of 992 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fa918 in realloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:85
    #1 0x79600bad37e3  (/home/adam/weechat/bin/weechat+0x48e7e3)
    #2 0x79600bad3898  (/home/adam/weechat/bin/weechat+0x48e898)
    #3 0x79600ef46d76  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d76)
    #4 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #5 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152
    #6 0x59a696cfa428 in plugin_init /home/adam/weechat-src/src/plugins/plugin.c:1409
    #7 0x59a696b1e004 in weechat_init /home/adam/weechat-src/src/core/weechat.c:675
    #8 0x59a696b1c55d in main /home/adam/weechat-src/src/gui/curses/normal/main.c:43
    #9 0x79601082a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #10 0x79601082a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #11 0x59a696b1c464 in _start (/home/adam/weechat/bin/weechat+0x9c464) (BuildId: 405e4255136b6fd0fefff49787a34a3e899b7b0c)

Direct leak of 768 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fa918 in realloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:85
    #1 0x79600b9e6a28  (<unknown module>)
    #2 0x79600b9e6cfc  (<unknown module>)
    #3 0x79600bb25de5  (/home/adam/weechat/bin/weechat+0x4e0de5)
    #4 0x79600b9ca49b  (<unknown module>)
    #5 0x79600ba5146a  (/home/adam/weechat/bin/weechat+0x40c46a)
    #6 0x79600baefe59  (/home/adam/weechat/bin/weechat+0x4aae59)
    #7 0x79600baf1e9d  (/home/adam/weechat/bin/weechat+0x4ace9d)
    #8 0x79600baf2024  (/home/adam/weechat/bin/weechat+0x4ad024)
    #9 0x79600baf327f  (/home/adam/weechat/bin/weechat+0x4ae27f)
    #10 0x79600ef46d7b  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d7b)
    #11 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #12 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152
    #13 0x59a696cfa428 in plugin_init /home/adam/weechat-src/src/plugins/plugin.c:1409
    #14 0x59a696b1e004 in weechat_init /home/adam/weechat-src/src/core/weechat.c:675
    #15 0x59a696b1c55d in main /home/adam/weechat-src/src/gui/curses/normal/main.c:43
    #16 0x79601082a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #17 0x79601082a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #18 0x59a696b1c464 in _start (/home/adam/weechat/bin/weechat+0x9c464) (BuildId: 405e4255136b6fd0fefff49787a34a3e899b7b0c)

Direct leak of 304 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fbb37 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x79600baebeb3  (/home/adam/weechat/bin/weechat+0x4a6eb3)
    #2 0x79600baf0b19  (/home/adam/weechat/bin/weechat+0x4abb19)
    #3 0x79600baf20c6  (/home/adam/weechat/bin/weechat+0x4ad0c6)
    #4 0x79600baf327f  (/home/adam/weechat/bin/weechat+0x4ae27f)
    #5 0x79600ef46d7b  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d7b)
    #6 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #7 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152
    #8 0x59a696cfa428 in plugin_init /home/adam/weechat-src/src/plugins/plugin.c:1409
    #9 0x59a696b1e004 in weechat_init /home/adam/weechat-src/src/core/weechat.c:675
    #10 0x59a696b1c55d in main /home/adam/weechat-src/src/gui/curses/normal/main.c:43
    #11 0x79601082a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #12 0x79601082a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #13 0x59a696b1c464 in _start (/home/adam/weechat/bin/weechat+0x9c464) (BuildId: 405e4255136b6fd0fefff49787a34a3e899b7b0c)

Direct leak of 80 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fbb37 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x79600b9e6173  (<unknown module>)
    #2 0x79600baebd82  (/home/adam/weechat/bin/weechat+0x4a6d82)
    #3 0x79600baf0b19  (/home/adam/weechat/bin/weechat+0x4abb19)
    #4 0x79600baf20c6  (/home/adam/weechat/bin/weechat+0x4ad0c6)
    #5 0x79600baf327f  (/home/adam/weechat/bin/weechat+0x4ae27f)
    #6 0x79600ef46d7b  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d7b)
    #7 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #8 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152
    #9 0x59a696cfa428 in plugin_init /home/adam/weechat-src/src/plugins/plugin.c:1409
    #10 0x59a696b1e004 in weechat_init /home/adam/weechat-src/src/core/weechat.c:675
    #11 0x59a696b1c55d in main /home/adam/weechat-src/src/gui/curses/normal/main.c:43
    #12 0x79601082a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #13 0x79601082a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #14 0x59a696b1c464 in _start (/home/adam/weechat/bin/weechat+0x9c464) (BuildId: 405e4255136b6fd0fefff49787a34a3e899b7b0c)

Direct leak of 68 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fbb37 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x79600b9e6173  (<unknown module>)
    #2 0x79600baebe18  (/home/adam/weechat/bin/weechat+0x4a6e18)
    #3 0x79600baf0b19  (/home/adam/weechat/bin/weechat+0x4abb19)
    #4 0x79600baf20c6  (/home/adam/weechat/bin/weechat+0x4ad0c6)
    #5 0x79600baf327f  (/home/adam/weechat/bin/weechat+0x4ae27f)
    #6 0x79600ef46d7b  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d7b)
    #7 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #8 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152
    #9 0x59a696cfa428 in plugin_init /home/adam/weechat-src/src/plugins/plugin.c:1409
    #10 0x59a696b1e004 in weechat_init /home/adam/weechat-src/src/core/weechat.c:675
    #11 0x59a696b1c55d in main /home/adam/weechat-src/src/gui/curses/normal/main.c:43
    #12 0x79601082a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #13 0x79601082a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #14 0x59a696b1c464 in _start (/home/adam/weechat/bin/weechat+0x9c464) (BuildId: 405e4255136b6fd0fefff49787a34a3e899b7b0c)

Direct leak of 32 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fbb37 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x79600b9e6173  (<unknown module>)
    #2 0x79600baebdb4  (/home/adam/weechat/bin/weechat+0x4a6db4)
    #3 0x79600baf0b19  (/home/adam/weechat/bin/weechat+0x4abb19)
    #4 0x79600baf20c6  (/home/adam/weechat/bin/weechat+0x4ad0c6)
    #5 0x79600baf327f  (/home/adam/weechat/bin/weechat+0x4ae27f)
    #6 0x79600ef46d7b  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d7b)
    #7 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #8 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152

Direct leak of 20 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fbb37 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x79600b9e6173  (<unknown module>)
    #2 0x79600baebd50  (/home/adam/weechat/bin/weechat+0x4a6d50)
    #3 0x79600baf0b19  (/home/adam/weechat/bin/weechat+0x4abb19)
    #4 0x79600baf20c6  (/home/adam/weechat/bin/weechat+0x4ad0c6)
    #5 0x79600baf327f  (/home/adam/weechat/bin/weechat+0x4ae27f)
    #6 0x79600ef46d7b  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d7b)
    #7 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #8 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152
    #9 0x59a696cfa428 in plugin_init /home/adam/weechat-src/src/plugins/plugin.c:1409
    #10 0x59a696b1e004 in weechat_init /home/adam/weechat-src/src/core/weechat.c:675
    #11 0x59a696b1c55d in main /home/adam/weechat-src/src/gui/curses/normal/main.c:43
    #12 0x79601082a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #13 0x79601082a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #14 0x59a696b1c464 in _start (/home/adam/weechat/bin/weechat+0x9c464) (BuildId: 405e4255136b6fd0fefff49787a34a3e899b7b0c)

Direct leak of 20 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fbb37 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x79600b9e6173  (<unknown module>)
    #2 0x79600baebd1e  (/home/adam/weechat/bin/weechat+0x4a6d1e)
    #3 0x79600baf0b19  (/home/adam/weechat/bin/weechat+0x4abb19)
    #4 0x79600baf20c6  (/home/adam/weechat/bin/weechat+0x4ad0c6)
    #5 0x79600baf327f  (/home/adam/weechat/bin/weechat+0x4ae27f)
    #6 0x79600ef46d7b  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d7b)
    #7 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #8 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152
    #9 0x59a696cfa428 in plugin_init /home/adam/weechat-src/src/plugins/plugin.c:1409
    #10 0x59a696b1e004 in weechat_init /home/adam/weechat-src/src/core/weechat.c:675
    #11 0x59a696b1c55d in main /home/adam/weechat-src/src/gui/curses/normal/main.c:43
    #12 0x79601082a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #13 0x79601082a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #14 0x59a696b1c464 in _start (/home/adam/weechat/bin/weechat+0x9c464) (BuildId: 405e4255136b6fd0fefff49787a34a3e899b7b0c)

Indirect leak of 131072 byte(s) in 1 object(s) allocated from:
    #0 0x7960110fb4d0 in calloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:77
    #1 0x79600b9e5af3  (<unknown module>)
    #2 0x79600b9e6819  (<unknown module>)
    #3 0x79600b9e6cfc  (<unknown module>)
    #4 0x79600bb25de5  (/home/adam/weechat/bin/weechat+0x4e0de5)
    #5 0x79600b9ca49b  (<unknown module>)
    #6 0x79600ba5146a  (/home/adam/weechat/bin/weechat+0x40c46a)
    #7 0x79600baefe59  (/home/adam/weechat/bin/weechat+0x4aae59)
    #8 0x79600baf1e9d  (/home/adam/weechat/bin/weechat+0x4ace9d)
    #9 0x79600baf2024  (/home/adam/weechat/bin/weechat+0x4ad024)
    #10 0x79600baf327f  (/home/adam/weechat/bin/weechat+0x4ae27f)
    #11 0x79600ef46d7b  (/usr/lib/debug/.build-id/98/2083caac6565fb72054d3529378d51bd903e8e.debug+0xe6d7b)
    #12 0x59a696cf3697 in plugin_call_init /home/adam/weechat-src/src/plugins/plugin.c:368
    #13 0x59a696cf9780 in plugin_auto_load /home/adam/weechat-src/src/plugins/plugin.c:1152
    #14 0x59a696cfa428 in plugin_init /home/adam/weechat-src/src/plugins/plugin.c:1409
    #15 0x59a696b1e004 in weechat_init /home/adam/weechat-src/src/core/weechat.c:675
    #16 0x59a696b1c55d in main /home/adam/weechat-src/src/gui/curses/normal/main.c:43
    #17 0x79601082a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #18 0x79601082a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #19 0x59a696b1c464 in _start (/home/adam/weechat/bin/weechat+0x9c464) (BuildId: 405e4255136b6fd0fefff49787a34a3e899b7b0c)

SUMMARY: AddressSanitizer: 395500 byte(s) leaked in 10 allocation(s).
```

### WeeChat version

4.3.1

### What OS are you using?

Ubuntu 24.04 LTS

### On which terminal are you running WeeChat?

xfce4-terminal

### Which terminal multiplexer are you using?

screen

**Comment by trygveaa** [comment]
Duplicate of #2046. It's an issue in Python 3.12, fixed in Python 3.13 (prerelease). Here's the Python issue: https://github.com/python/cpython/issues/116510

**Comment by adsr** [comment]
Apologies for the dupe. Thank you!
